### PR TITLE
enhancement(observability): Convert `EventsSent` to a registered event

### PR DESCRIPTION
**Pull request by bruceg**
Ref #13691

### CODE REVIEW

File: src/topology/builder.rs [hunk 1]
```diff
@@ -822,14 +822,12 @@ fn build_task_transform(
                 byte_size: events.size_of(),
             })
         });
+    // FIXME: Does this get the tags right (span)?
```
Comment (neuronull) [inline src/topology/builder.rs:825]: Still being looked into? Does it need a follow-up issue ?

File: lib/vector-common/src/internal_event/mod.rs [hunk 2]
```diff
@@ -104,6 +106,10 @@ pub struct Count(pub usize);
 #[derive(Clone, Copy)]
 pub struct CountByteSize(pub usize, pub usize);
 
+// Wrapper types used to hold inputs for registering an event
+
+pub struct Output(pub Option<SharedString>);
```
Comment (neuronull) [inline lib/vector-common/src/internal_event/mod.rs:111]: It is probably just my lack of understanding but thought it was odd to see `Output` below a comment talking about inputs.

**Review by neuronull — COMMENTED** [review]
LGTM.

Something to consider- perhaps documenting as a function or trait comment somewhere, describing the design and giving an example of what the typical or common usage of `EventSent` would look like. Or even for `RegisterInternalEvent` in general.

That could be helpful for the developer/new contributor experience.

File: lib/vector-common/src/internal_event/mod.rs (hunk 2 above)
Comment (bruceg) [reply lib/vector-common/src/internal_event/mod.rs:111]: The registered event uses a source or transform output name as a parameter for registration. I'll reword to clarify.

**Review by neuronull — APPROVED** [review]

**Review by spencergilbert — APPROVED** [review]

File: lib/vector-common/src/internal_event/events_sent.rs [hunk 3]
```diff
@@ -1,34 +1,38 @@
...
 
 pub const DEFAULT_OUTPUT: &str = "_default";
 
 #[derive(Debug)]
-pub struct EventsSent<'a> {
-    pub count: usize,
-    pub byte_size: usize,
-    pub output: Option<&'a str>,
+pub struct EventsSent {
+    pub output: Option<SharedString>,
```
[the remark this answers is not in the document]
Comment (StephenWakely) [reply lib/vector-common/src/internal_event/events_sent.rs:10]: From [here](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#componenteventssent):

> output - OPTIONAL, for components that can use multiple outputs, the name of the output that events were sent to. For events sent to the default output, this value MUST be _default.

**Comment by bruceg** [comment]
This is running into an issue where all of the component compliance tests for the `http_server` source are failing. None of the other sources appear to be having problems, and this PR doesn't actually touch sources anyways (only `SourceSender`). I have no idea at this point why it's not working, and no easy ideas how to move it forward.

**Pull request merged**